plugin/builtin/sortordertransformer/SortOrderTransformer.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+//go:generate pluginator
+package main
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/kustomize/api/resmap"
+	"sigs.k8s.io/kustomize/api/resource"
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/resid"
+	"sigs.k8s.io/yaml"
+)
+
+// Sort the resources using a customizable ordering based of Kind.
+// Defaults to the ordering of the GVK struct, which puts cluster-wide basic
+// resources with no dependencies (like Namespace, StorageClass, etc.) first,
+// and resources with a high number of dependencies
+// (like ValidatingWebhookConfiguration) last.
+type plugin struct {
+	SortOptions *types.SortOptions `json:"sortOptions,omitempty" yaml:"sortOptions,omitempty"`
+	// TODO: We can't actually be 100% sure that this isn't the top-level
+	// kustomization, as that information is not propagated through
+	// PluginHelpers. Why not? Because propagating this information would
+	// require extensive code changes and changes to the public API, so we opt
+	// to start small. For now, the transformer will only be called explicitly
+	// if the appropriate fields are set in the kustomization.yaml file, which
+	// will also set the IsTopLevel field.
+	//
+	// See: https://github.com/kubernetes-sigs/kustomize/pull/4019#issuecomment-1159771417
+	IsTopLevel bool
```

Comment:
As it stands today, transformers aren't aware of the structure around which they are invoked, and I would prefer to keep it that way. The type fields in structs like this one should describe the work the transformer itself does, and if we were to reframe this one to that end, it would be something like `DoNothing bool`--odd IMO. If we want to prevent the transformer from being used directly by end users (i.e. via `sortOptions` and `transformers` fields in arbitrary Kustomizations), I think we should accomplish that by not configuring it in built-ins in the first place, not by making it no-op. I think that's possible by excluding it from a few places in `api/internal/target/kusttarget_configplugin.go`, which appears to have been done for the legacy transformer. If I'm wrong about that, let's reconsider restricting this to the top level.

Either way, I'd like to see tests demonstrating the behaviour when a child kustomization contains the field, and when the top-level kustomization attempts to invoke the transformer via the `transformers` field directly.

Reply:
Not wiring the transformer through `api/internal/target/kusttarget_configplugin.go` seems reasonable. I'll try that!

> Either way, I'd like to see tests demonstrating the behaviour when a child kustomization contains the field, and when the top-level kustomization attempts to invoke the transformer via the transformers field directly.

Ack! Seems reasonable!